### PR TITLE
feat: add changelog-generator

### PR DESCRIPTION
**Pull request by bill-mahoney**
Periodic Jenkins job to automatically generate CHANGELOG.md files each EdgeX Foundry repository.

- Iterates through **non-archived** repositories
- Clones each repository in parallel.
- Copies CHANGELOG.tpl.md and config.yml to each repository (If one already exists in the target repository it will not overwrite it).
- Locally deletes `dev` tags.
- Generates and archives the CHANGELOG.md for each repository.

Contributing community members can download this changelog for their repository and use it as a basis for a PR to update the changelog prior to release.

This job utilizes [git-chglog](https://github.com/git-chglog/git-chglog) for CHANGELOG.md generation.

Pipeline View:
image

Functional Testing / Sandbox Links
[Build Console](https://jenkins.edgexfoundry.org/sandbox/job/Functional-Testing/job/changelog-generator-scm/24/)
[Build Artifacts](https://jenkins.edgexfoundry.org/sandbox/job/Functional-Testing/job/changelog-generator-scm/24/artifact/)

Custom Query: <blank>  (Do all unarchived repos) - [Pipeline](https://jenkins.edgexfoundry.org/sandbox/blue/organizations/jenkins/Functional-Testing%2Fchangelog-generator-scm/detail/changelog-generator-scm/26/pipeline)
Custom Query: "device" (Do all device services) - [Pipeline](https://jenkins.edgexfoundry.org/sandbox/blue/organizations/jenkins/Functional-Testing%2Fchangelog-generator-scm/detail/changelog-generator-scm/27/pipeline)

[Sample Output](https://github.com/bill-mahoney/changelog-automation/tree/main/Samples)

image

Signed-off-by: Bill Mahoney <bill.mahoney@intel.com>

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    triggers {
+        cron '@weekly'
+    }
+    stages {
+        stage('Parallel Changelog Generator'){
+            steps {
+                script {
+                    jobs = sh (
+                        script: 'curl -i "https://api.github.com/search/repositories?q=user:edgexfoundry+archived:false&per_page=200" | grep clone_url',
```
Comment (soda480) [inline Jenkinsfile:52]: I wonder if we could parameterize the query so a user can execute the change log automation on a select set of repositories that match a criteria.  Its a nice to have, not sure if it is necessary.

File: .chglog-default/CHANGELOG.tpl.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+{{ end -}}
+
+{{- if .Versions }}
+[Unreleased]: {{ .Info.RepositoryURL }}/compare/{{ $latest := index .Versions 0 }}{{ $latest.Tag.Name }}...HEAD
+{{ range .Versions -}}
+{{ if .Tag.Previous -}}
+[{{ .Tag.Name }}]: {{ $.Info.RepositoryURL }}/compare/{{ .Tag.Previous.Name }}...{{ .Tag.Name }}
+{{ end -}}
+{{ end -}}
+{{ end -}}
```
Comment (soda480) [inline .chglog-default/CHANGELOG.tpl.md:36]: holy smokes, will need to come up to speed on what all this means. :)

**Review by soda480 — APPROVED** [review]
Just a few comments, but looks good.

File: .chglog-default/CHANGELOG.tpl.md (hunk 2 above)
Comment (ernestojeda) [reply .chglog-default/CHANGELOG.tpl.md:36]: This is Go templates https://blog.gopheracademy.com/advent-2017/using-go-templates/.

File: Jenkinsfile (hunk 1 above)
Comment (ernestojeda) [reply Jenkinsfile:52]: Would be nice to have for sure.

**Review by ernestojeda — APPROVED** [review]
LGTM

**Review by cjoyv — APPROVED** [review]
Looks great 👍

File: .chglog-default/CHANGELOG.tpl.md (hunk 2 above)
Comment (bill-mahoney) [reply .chglog-default/CHANGELOG.tpl.md:36]: First time here too, I had the same reaction :)   Once I dug into it, it was kind of fun.

File: Jenkinsfile (hunk 1 above)
Comment (bill-mahoney) [reply Jenkinsfile]: Thanks for the inspiration, I added it as a parameter 👍

**Pull request merged**